packages/components/src/responsive-toolbar-group/index.tsx

```diff
@@ -0,0 +1,215 @@
+import { ToolbarGroup, ToolbarButton, Dropdown, MenuItem, MenuGroup } from '@wordpress/components';
+import { Icon, chevronDown } from '@wordpress/icons';
+import classnames from 'classnames';
+import { ReactChild, useCallback, useEffect, useRef, useState } from 'react';
+
+import './style.scss';
+
+interface GroupedIndexStore {
+	[ key: string ]: boolean;
+}
+
+const ResponsiveToolbarGroup = ( {
+	children,
+	className = '',
+	hideRatio = 0.99,
+	showRatio = 1,
+	rootMargin = '0px',
+	onClick = () => null,
+	initialActiveIndex = -1,
+}: {
+	children: ReactChild[];
+	className?: string;
+	hideRatio?: number;
+	showRatio?: number;
+	rootMargin?: string;
+	onClick?: ( index: number ) => void;
+	initialActiveIndex?: number;
+} ) => {
+	const classes = classnames( 'responsive-toolbar-group', className );
+
+	const containerRef = useRef< HTMLDivElement >( null );
+	const [ calculatedOnce, setCalculatedOnce ] = useState< boolean >( false );
+	const [ activeIndex, setActiveIndex ] = useState< number >( initialActiveIndex );
+	const [ groupedIndexes, setGroupedIndexes ] = useState< GroupedIndexStore >( {} );
+	const { current: shadowListItems } = useRef< HTMLButtonElement[] >( [] );
+
+	const assignRef = ( index: number, element: HTMLButtonElement ) => {
+		shadowListItems[ index ] = element;
+	};
+
+	const getChildrenToRender = () =>
+		Object.keys( groupedIndexes ).map( ( index ) => ( {
+			index,
+			grouped: groupedIndexes[ index ],
+			child: children[ parseInt( index ) ],
+		} ) );
```

Comment:
Can this or any other function be moved outside the function component? I don't see any dependency and we would avoid declaring them again on every render.

Reply:
> 1. On smaller viewports, I think it would be preferable having a scrolling menu?

I was thinking on implementing this one on a different PR since this one is getting to big.

> 2. Let's make sure that the navigation header is always on top

Will do this one.

> 3. On some viewports there is padding missing

Will do this one as well.

> 4. I really like the animation on selection 🎉 ! Have you considered adding a hover effect too?

Thanks the animation comes from `@wordpress/components/ToolbarGroup` which I reused here, I can add a hover through css but I wonder if we should also add a commit upstream.

> 5. For the future:
>    a) should we make this menu sticky?
>    b) can we avoid re-rendering the whole page to avoid the scroll-top?

Good follow-ups